mongomock/collection.py

```diff
@@ -1570,8 +1570,13 @@ def _extend_collection(out_collection, field, expression):
                                     elif operator == "$last":
                                         doc_dict[field] = values[-1]
                                     elif operator == "$addToSet":
+                                        doc_dict[field] = []
                                         val_it = (val or None for val in values)
-                                        doc_dict[field] = set(val_it)
+                                        # don't use set in case elt in not hashable (like dicts)
+                                        for elt in val_it:
+                                            if elt not in doc_dict[field]:
+                                                doc_dict[field].append(elt)
+                                        doc_dict[field].reverse()
```

Comment:
I'm not sure why you are reversing it: according to the [doc](https://docs.mongodb.com/manual/reference/operator/aggregation/addToSet/), the order of the elements in the output array is unspecified.

Reply:
Actually, if you remove this line, the test fails. It seems the order may not be guarantied but this must be how it is implemented. I can remove this line and update the test (check that ensembles are equal and so don't take the order into account). WDYT ?

Reply:
Exactly, I like your second solution: update the test to match the expectation of MongoDB instead of having extra expectations for Mongomock.